### PR TITLE
fix(memory): disambiguate replace when new_content already exists

### LINKED ISSUE #60089: [Bug]: memory replace → "Multiple entries matched" → fallback add creates duplicate entries with no signal
### Bug Description

`MemoryStore.replace()` refuses an ambiguous replace when `old_text` substring-matches two or more entries ("Multiple entries matched"), which is correct on its own. But when the model's intended replacement text is *already present verbatim* as a separate entry, that generic error sets a trap: the model's natural fallback is `add()`, which then creates a duplicate of the entry that already said what it wanted to write. The model never gets a signal that the fact it wanted to record already exists.

`add()` already guards the simpler case - exact duplicates are rejected with `"Entry already exists (no duplicate added)"`, but the replace path gives no equivalent hint, so the duplicate arrives via the replace → add two-step.

Observed as a repeatable tool-misuse pattern running a local model (qwen3.6-35b) against the file memory store: replace → ambiguous match → fallback add → duplicate entry.

### Steps to Reproduce

1. Seed the file memory store with two entries sharing a substring (e.g. both containing "prefers dark mode") and a third entry whose text is exactly what the model would write as the update.
2. Call `memory(action="replace", old_text="prefers dark mode", new_content=<the third entry's exact text>)`.
3. Get the generic "Multiple entries matched" error.
4. Call `memory(action="add", content=<same text>)` with content differing even trivially from the existing entry (whitespace, punctuation), a duplicate is stored.

### Expected Behavior

When an ambiguous replace's `new_content` already exists verbatim as a separate entry, the error should say so and point at a batch `operations` call to disambiguate instead of the generic ambiguity message, that invites a duplicate-creating `add()`.

### Actual Behavior

Generic "Multiple entries matched" error, then a duplicate entry after the fallback add.

### Affected Component

Agent Core (conversation loop, context compression, memory)

### Messaging Platform (if gateway-related)

_No response_

### Debug Report

```shell
Not attaching "hermes debug share" links for this one: it's a code-level behavior
gap in tools/memory_tool.py, fully reproducible from unit tests and not
environment-specific, and my recent logs contain private gateway-session content.
Confirmed the gap still exists on current main by inspection — replace() returns
the generic ambiguity error with no check against existing entry content. Happy
to share specific redacted log excerpts on request.

System info (from "hermes debug share --local"):

version:      0.18.0 (2026.7.1)
os:           Darwin 25.4.0 arm64 (macOS 15, Apple Silicon)
python:       3.11.15
openai_sdk:   2.24.0
provider:     lmstudio (local OpenAI-compatible endpoint)
model:        Qwen3.6-35B-A3B-MLX-8bit
toolsets:     hermes-cli
memory:       built-in
gateway:      telegram
```

### Operating System

macOS 15

### Python Version

3.11.15

### Hermes Version

0.18.0

### Additional Logs / Traceback (optional)

```shell

```

### Root Cause Analysis (optional)

_No response_

### Proposed Fix (optional)

_No response_

### Are you willing to submit a PR for this?

- [x] I'd like to fix this myself and submit a PR

### PR DESCRIPTION
**Pull request by Bartok9**
## Summary
Closes #60089. When `memory` replace matches multiple entries and `new_content` already exists as its own entry, return an explicit already-exists error instead of generic ambiguity (which steered models into duplicate `add()`).

## Verification
`python3 -m pytest tests/tools/test_memory_tool.py -q -k replace_ambiguous`

### CODE REVIEW

File: tools/memory_tool.py [hunk 1]
```diff
@@ -481,6 +481,22 @@ def replace(self, target: str, old_text: str, new_content: str) -> Dict[str, Any
                 # If all matches are identical (exact duplicates), operate on the first one
                 unique_texts = {e for _, e in matches}
                 if len(unique_texts) > 1:
+                    # Ambiguous substring match traps models into add() as a
+                    # fallback. If new_content already exists verbatim as its
+                    # own entry, say so explicitly so the agent does not create
+                    # a near-duplicate (#60089).
+                    if new_content in entries:
```
Comment (teknium1) [inline tools/memory_tool.py]: This also matches when `new_content` equals one of the ambiguous candidates in `unique_texts`, despite the comment promising an entry of its own. Require `new_content not in unique_texts` as well, and add a regression test for a matched candidate so a legitimate ambiguous replace is not incorrectly reported as a no-op.

File: tools/memory_tool.py [hunk 2]
```diff
@@ -481,6 +481,22 @@ def replace(self, target: str, old_text: str, new_content: str) -> Dict[str, Any
...
+                        return {
+                            "success": False,
+                            "error": (
+                                f"Multiple entries matched '{old_text}', and "
+                                f"new_content already exists as its own entry "
+                                f"(no change needed). Prefer operations=[] to "
+                                f"disambiguate a replace, or skip the write."
```
Comment (teknium1) [inline tools/memory_tool.py]: `operations=[]` is not a valid remediation: `apply_batch()` rejects an empty list at current-main `tools/memory_tool.py:575`, and a non-empty batch replace retains the same distinct-match rejection at line 624. Please direct the caller to skip the write, use a more specific `old_text`, or make explicit batch removals after identifying the target entries.

**Review by teknium1 — COMMENTED** [review]
Thanks for addressing a real current-main gap: `tools/memory_tool.py:480-489` still returns only the generic ambiguity response.

### Problems
- `tools/memory_tool.py:488` checks only `new_content in entries`. That is also true when `new_content` is one of `unique_texts`, so the new response mislabels an ambiguous candidate as a separate entry and says no change is needed.
- `tools/memory_tool.py:495` recommends `operations=[]`, but `apply_batch()` rejects empty operations at `tools/memory_tool.py:575`; a non-empty batch replace with the same ambiguous text also fails at `tools/memory_tool.py:624-628`.
- The new test covers only the separate-entry case. Add the matched-candidate negative case so this distinction is protected.

### Suggested changes
- Require `new_content not in unique_texts`, add the paired regression test, and replace the empty-batch instruction with valid disambiguation guidance.
- Please keep the unrelated empty-argument test additions out of this focused fix; equivalent tests were removed in `39975613b1`.

This is an automated hermes-sweeper review.

**Comment by Bartok9** [comment]
Review fix: separate-entry already-exists now requires `new_content not in unique_texts` (matched candidates no longer mislabeled). Error no longer recommends `operations=[]` (apply_batch rejects empty lists) and asks for a more specific `old_text`. Added matched-candidate negative test. Local: `pytest tests/tools/test_memory_tool.py::TestMemoryStoreReplace` + ruff clean.

**Comment by GottZ** [comment]
> *This was generated by AI during triage.*

#### Summary

Nine PRs address #60089's ambiguous-replace feedback gap. They all attempt to detect an already-stored `new_content`, but only #60147, #66916, and the revised #71819 distinguish a genuinely separate entry from one of the ambiguous match candidates; several others check only the matched entries or misclassify matched candidates.

#### Related pull requests

- #60090 **[closed]** `partial` — (+80/-0) — closed reference implementation: adds the replace-path signal, dispatcher pre-validation, and tests, but `new_content in entries` also matches an ambiguous candidate and therefore does not fully distinguish the reported separate-entry cause. The automated close verdict cites the existing exact-add guard at `tools/memory_tool.py:347-361` and its test at `tests/tools/test_memory_tool.py:288-292`; this PR remains relevant as the source of the duplicated two-layer approach, not as a complete fix for the near-duplicate behavior.
- #60093 **[closed]** `related` — (+36/-2) — closed duplicate of #60090: changes `replace()` and batch messaging, but searches only the ambiguous matches, so it misses the issue's defining case where the existing replacement is a separate non-matching entry.
- #60147 **[closed]** `partial` — (+16/-0) — closed duplicate of #60090: its `new_content in entries and new_content not in unique_texts` condition correctly models the separate-entry cause, but it contains only the store-level guard and no regression tests.
- #60208 **[closed]** `related` — (+11/-0) — closed duplicate of #60090: checks whether `new_content` equals one of the matched entries, which is specifically not the separate third-entry scenario reported by #60089 and can mislabel a legitimate ambiguous candidate.
- #60216 **[closed]** `related` — (+36/-2) — closed duplicate of #60090: extends the same matched-entry check into batch handling, but still does not detect replacement content stored as a separate non-matching entry.
- #60321 **[closed]** `partial` — (+80/-0) — closed duplicate of #60090: it is a cherry-pick of #60090 and therefore preserves both that PR's tests and dispatcher guard, as well as its failure to exclude `new_content` that is itself an ambiguous candidate.
- #60511 **[closed]** `related` — (+17/-0) — closed duplicate of #60090: checks `new_content in unique_texts`, so it handles content among the ambiguous candidates rather than the separate-entry collision that triggers the reported replace-to-add loop.
- #66916 **[closed]** `best fix` — (+67/-0) — closed as duplicate of #60208 but still relevant as a recorded best-fix reference: unlike #60208, its diff correctly requires `new_content in entries and new_content not in unique_texts`, exposes `existing_duplicate`, and tests both the separate-entry case and the matched-candidate negative case.
- #71819 `best fix` — (+69/-0) — keep open with a salvage path: the revised diff addresses the keep_open review on #71819 by excluding `unique_texts`, removing the invalid `operations=[]` guidance, and adding the requested matched-candidate negative test. Retain that focused guard and its paired regression tests, but remove the unrelated empty-argument tests identified by the contributor review before further consideration.

#### Duplicates

#60321 is a direct cherry-pick of #60090. #60093 and #60216 carry the same store-and-batch patch; #60208 and #60511 are narrower variants of that matched-entry logic. At the issue level, #60090, #60093, #60147, #60208, #60216, #60321, #60511, #66916, and #71819 all pursue the same targeted ambiguous-replace signal, with #66916 and revised #71819 providing the strongest equivalent separate-entry distinction.

#### Suggested consolidation

Keep #71819 open with a salvage path: preserve its corrected `new_content in entries and new_content not in unique_texts` guard and the positive/negative regression pair, then remove the unrelated empty-argument tests as requested in the contributor review. This is consistent with the visible keep_open review and with the recorded best-fix status of #71819; it also incorporates the key distinction demonstrated by recorded best-fix #66916 without recommending that closed PR be reopened. The already-closed #60090, #60093, #60147, #60208, #60216, #60321, #60511, and #66916 can remain closed as duplicate or historical reference implementations.

#### Complex graph

```mermaid
flowchart TD
    classDef open fill:#dbeafe,stroke:#1d4ed8,color:#1e3a8a
    classDef merged fill:#dcfce7,stroke:#15803d,color:#14532d
    classDef closed fill:#e5e7eb,stroke:#6b7280,color:#1f2937
    classDef unverified fill:#f3f4f6,stroke:#9ca3af,color:#374151
    classDef best stroke-width:3px,stroke:#b45309
    classDef target stroke-width:3px,stroke:#4338ca
    I60089(["issue #60089 (open)"])
    subgraph Dup60090 ["PRs duplicating each other"]
        P60090["PR #60090 (closed)"]
        P60147["PR #60147 (closed)"]
        P60321["PR #60321 (closed)"]
        P66916["PR #66916 (closed)"]
        P71819["PR #71819 (open)"]
    end
    P71819 -->|best fix| I60089
    class I60089 open
    class P60090 closed
    class P60147 closed
    class P60321 closed
    class P66916 closed
    class P71819 open
    class P66916 best
    class P71819 best
    class P71819 target
    click I60089 "https://github.com/NousResearch/hermes-agent/issues/60089"
    click P60090 "https://github.com/NousResearch/hermes-agent/pull/60090"
    click P60147 "https://github.com/NousResearch/hermes-agent/pull/60147"
    click P60321 "https://github.com/NousResearch/hermes-agent/pull/60321"
    click P66916 "https://github.com/NousResearch/hermes-agent/pull/66916"
    click P71819 "https://github.com/NousResearch/hermes-agent/pull/71819"
```
*Graph: solid arrow = fixes / best fix, dashed arrow = partial or unverified (see edge label); boxed group = PRs duplicating each other; amber border = best fix; indigo border = target; gray node = closed (state tag in the node label).*

_Cross-PR triage: Reviewed 9 pull requests and 1 issue in this complex. Each diff was read against this issue; Assessment working set: 35 kB of PR diffs, 16 kB of issue/PR text, 6 kB of discussion (19 comments), 16 verify verdicts. verdicts reflect diff content, not PR titles. Part of an automated triage batch._

**Comment by Bartok9** [comment]
Thanks for the consolidation triage — agreed this is the keep-open salvage path.

Pushed a follow-up that drops the unrelated empty old/new replace tests so the PR stays focused on the separate-entry already-exists guard + the paired matched-candidate negative regression (and the invalid `operations=[]` guidance was already removed).
